api/services/bulk-docs.js

```diff
@@ -35,35 +34,69 @@ const generateBatches = (docs, batchSize) => {
   return batches;
 };
 
-const generateBatchPromise = (batch, res, { isFinal } = {}) => {
-  return db.medic.bulkDocs(batch).then(result => {
-    let resString = JSON.stringify(result);
-    resString += isFinal ? '' : ',';
-    res.write(resString);
-  });
+const getDocsToModify = ids => {
+  let docsToDelete;
+  return db.medic.allDocs({ keys: ids, include_docs: true })
+    .then(result => {
+      docsToDelete = markAsDeleted(result);
+      return utils.updateParentContacts(docsToDelete);
+    })
+    .then(docsToUpdate => {
+      const allDocs = docsToDelete.concat(docsToUpdate);
+      checkForDuplicates(allDocs);
+      return allDocs;
+    });
 };
 
-module.exports = {
-  bulkDelete: (docs, res, { batchSize = 100 } = {}) => {
-    let docsToDelete;
-    const keys = docs.map(doc => doc._id);
-    return db.medic.allDocs({ keys, include_docs: true })
-      .then(result => {
-        docsToDelete = markAsDeleted(result);
-        return utils.updateParentContacts(docsToDelete);
+const deleteBatches = (batches, res, docsToRetry, attemptNumber) => {
```

Comment:
It also seems a bit weird to pass the attempt number in just to work out whether you should put pipe out a `,`